semantic-tags/src/Tags/Tagging/Precise.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE AllowAmbiguousTypes, ConstraintKinds, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, RankNTypes, ScopedTypeVariables, TypeApplications, TypeOperators #-}
+module Tags.Tagging.Precise
+( Tags
+, ToTags(..)
+, yield
+, runTagging
+, GFoldable1(..)
+) where
+
+import Control.Effect.Pure
+import Control.Effect.Reader
+import Control.Effect.Writer
+import Data.Monoid (Endo(..))
+import GHC.Generics
+import Prelude hiding (span)
+import Source.Loc (Loc)
+import Source.Span
+import Source.Source
+import Tags.Tag
+
+type Tags = Endo [Tag]
+
+class ToTags t where
+  tags :: Source -> t Loc -> [Tag]
+
+
+yield :: (Carrier sig m, Member (Writer Tags) sig) => Tag -> m ()
+yield = tell . Endo . (:) . modSpan toOneIndexed where
+  modSpan f t@Tag{ span = s } = t { span = f s }
+  toOneIndexed (Span (Pos l1 c1) (Pos l2 c2)) = Span (Pos (l1 + 1) (c1 + 1)) (Pos (l2 + 1) (c2 + 1))
+
+runTagging :: Source -> ReaderC Source (WriterC Tags PureC) () -> [Tag]
+runTagging source
+  = ($ [])
+  . appEndo
+  . run
+  . execWriter
+  . runReader source
+
+
+-- FIXME: move GFoldable1 into semantic-ast.
+class GFoldable1 c t where
```

Comment:
cc @aymannadeem—I think we should move functionality like this, and an `Element` class for projecting out parts of `:+:`-constructed trees, into `semantic-ast`, once it’s ready.

Reply:
Note that unlike a lot of generic patterns, there’s no non-`G`-prefixed version of this class. It basically just wasn’t very useful since you’d have to define instances of it for each type, which is kind of what we’re trying to avoid in the first place.